### PR TITLE
feat: Workflow Intelligence — smart status, phase tracking, naming clarity (#72)

### PR DESCRIPTION
**Pull request by harshanandak**
## Summary

- **Smart status script** (`scripts/smart-status.sh`): Ranks all open/in-progress/blocked issues by composite score (priority × unblock chain × type × status boost × epic proximity × staleness), grouped into actionable categories (Resume → Unblock Chains → Ready Work → Blocked → Backlog)
- **Session detection**: Parses `git worktree list --porcelain` to show active sessions, maps branches to beads issues
- **Two-tier conflict detection**: Tier 1 (file-level overlap via `git diff`) + Tier 2 (actual merge conflicts via `git merge-tree`, git 2.38+)
- **Phase tracking**: `/plan` now creates epic at Phase 1 entry with stage transitions at each phase boundary
- **Naming clarity**: `forge-validate` renamed to `forge-preflight`; disambiguation note added to `/validate` command
- **Dynamic commands rule**: Added to AGENTS.md and CLAUDE.md — never hardcode example output in command files

## Design Doc
See: [docs/plans/2026-03-21-workflow-intelligence-design.md](docs/plans/2026-03-21-workflow-intelligence-design.md)

## Decisions Log
See: [docs/plans/2026-03-21-workflow-intelligence-decisions.md](docs/plans/2026-03-21-workflow-intelligence-decisions.md)

## Beads Issues
Closes: forge-68oj (epic), forge-vhxt, forge-mwxb, forge-0xic, forge-w69s

## Key Decisions
1. **Single jq invocation** for scoring — sub-second for 87+ issues, no per-issue subprocess spawns
2. **Taskwarrior-inspired formula** with multiplicative factors (not additive) including epic proximity and tiered staleness
3. **Two-tier conflict detection** — file overlap (fast, always) + merge-tree (accurate, git 2.38+)
4. **Branch-name matching** for session detection — no new beads metadata needed
5. **Corrected rename scope** — only `bin/forge-validate.js` and package.json renamed; `/validate` command, `lib/commands/validate.js`, and `scripts/validate.sh` kept unchanged

## TDD Test Coverage
- Smart status scoring: 16 tests (all factors, edge cases, error paths)
- Smart status grouping: 10 tests (all categories, formatting, NO_COLOR)
- Session detection: 7 tests (matching, orphan branches, JSON mode)
- Conflict detection: 6 tests (overlap, truncation, annotations)
- Tier 2 merge-tree: 5 tests (real conflicts, clean merges, version fallback)
- Preflight rename: 7 tests (existence, absence, docs)
- Disambiguation note: 5 tests (content, agent sync)
- Status command: 4 tests (smart-status reference, agent sync)
- Plan phase tracking: 7 tests (early creation, stage transitions)
- **Total: 67 new tests, 0 regressions**

## Security Review
- OWASP Top 10: A03 Injection is the only material risk — mitigated via `sanitize()` function and `--` separator in all git commands
- No new dependencies (jq required, available on all platforms)
- `set -euo pipefail`, all variables quoted, `printf '%s'` over `echo`
- npm audit: 0 vulnerabilities

## Test Plan
- [x] Type check passing (no TypeScript)
- [x] Lint passing (0 errors, 0 warnings, --max-warnings 0)
- [x] Tests passing (1781 pass, 31 skip, 1 pre-existing fail on master)
- [x] Security scan: 0 vulnerabilities
- [x] Rebased on latest master (2 new PRs merged)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Confidence Score: 4/5</h3></summary>

- PR is safe to merge with one minor inconsistency to address in a follow-up
- The core smart-status engine is well-tested (67 new tests, 0 regressions) and a thorough prior review cycle addressed all the significant logic bugs. The single new issue — `validateDev` accepting a narrower set of branch prefixes than `smart-status.sh` — is a usability inconsistency rather than a data-loss or security risk, and the forge-preflight binary is new in this PR so there's no regression risk.
- bin/forge-preflight.js — the `validateDev` branch prefix allowlist should be expanded to match the generic slug handling now present in smart-status.sh
</details>

<details open><summary><h3>Comments Outside Diff (5)</h3></summary>

1. `bin/forge-preflight.js`, line 130-141 ([link](https://github.com/harshanandak/forge/blob/73d2898d17787e0eab26ccfddbac7ae1a41030e4/bin/forge-preflight.js#L130-L141)) 

   <a href="#">P1</a> **`validateDev` checks wrong plans directory**

   `validateDev` looks for plan files in `.claude/plans/`, but the `/plan` command writes design documents to `docs/plans/` (e.g., `docs/plans/2026-03-21-workflow-intelligence-design.md`). As a result, `forge-preflight dev` will always report "No plan file found" even when a developer has correctly run `/plan` and produced a plan under `docs/plans/`.

   

   Note: this was present in the original `forge-validate.js` — it is not newly introduced by the rename — but the PR brings it into scope for review and it will cause false failures for users running `forge-preflight dev`.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge-preflight.js
   Line: 130-141

   Comment:
   **`validateDev` checks wrong plans directory**

   `validateDev` looks for plan files in `.claude/plans/`, but the `/plan` command writes design documents to `docs/plans/` (e.g., `docs/plans/2026-03-21-workflow-intelligence-design.md`). As a result, `forge-preflight dev` will always report "No plan file found" even when a developer has correctly run `/plan` and produced a plan under `docs/plans/`.

   

   Note: this was present in the original `forge-validate.js` — it is not newly introduced by the rename — but the PR brings it into scope for review and it will cause false failures for users running `forge-preflight dev`.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `bin/forge-preflight.js`, line 185 ([link](https://github.com/harshanandak/forge/blob/3a1565c0a873b131887353dd867633587822f39f/bin/forge-preflight.js#L185)) 

   <a href="#">P1</a> **`fs.readdirSync({ recursive })` silently fails on Node < 18.17**

   The `{ recursive: true }` option for `fs.readdirSync` was introduced in Node.js 18.17.0. On any Node.js version below that (e.g., Node 16, or Node 18.0–18.16), calling this throws a `TypeError: The "options.recursive" property is not supported`. Because the call is inside `try { ... } catch { return false; }`, the error is silently swallowed, causing the "Tests exist" check to always report `false` — i.e., "No test files found. Write tests before shipping!" — even when test files are present.

   There is no `engines` field in `package.json` pinning a minimum Node version, so this could affect any consumer still on an older Node.

   A compatible fallback that works on Node 12+:

   ```javascript
   const files = fs.readdirSync(dir);
   return files.some(
     (f) => f.includes('.test.') || f.includes('.spec.'),
   );
   ```

   This only searches one level deep rather than recursively, but test files are typically at the top of the `test/` directory anyway. If recursive search is truly needed, a small manual recursive helper or a dependency like `glob` would work portably.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge-preflight.js
   Line: 185

   Comment:
   **`fs.readdirSync({ recursive })` silently fails on Node < 18.17**

   The `{ recursive: true }` option for `fs.readdirSync` was introduced in Node.js 18.17.0. On any Node.js version below that (e.g., Node 16, or Node 18.0–18.16), calling this throws a `TypeError: The "options.recursive" property is not supported`. Because the call is inside `try { ... } catch { return false; }`, the error is silently swallowed, causing the "Tests exist" check to always report `false` — i.e., "No test files found. Write tests before shipping!" — even when test files are present.

   There is no `engines` field in `package.json` pinning a minimum Node version, so this could affect any consumer still on an older Node.

   A compatible fallback that works on Node 12+:

   ```javascript
   const files = fs.readdirSync(dir);
   return files.some(
     (f) => f.includes('.test.') || f.includes('.spec.'),
   );
   ```

   This only searches one level deep rather than recursively, but test files are typically at the top of the `test/` directory anyway. If recursive search is truly needed, a small manual recursive helper or a dependency like `glob` would work portably.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

3. `bin/forge-preflight.js`, line 219-224 ([link](https://github.com/harshanandak/forge/blob/1d6fbefc5efbf047dc35ca33f98b80ff8e0b1ef4/bin/forge-preflight.js#L219-L224)) 

   <a href="#">P2</a> **`npm test` output silently discarded on failure**

   `execFileSync("npm", ["test"], { stdio: "pipe" })` captures and then discards all stdout/stderr from the test run. When this check fails, `forge-preflight ship` reports only:

   ```
   ✗ Tests pass
       Tests are failing. Fix them before shipping: npm test
   ```

   The user gets no indication of _which_ tests failed or _why_ — they must re-run `npm test` manually to see the output. This defeats part of the purpose of a preflight check that's meant to surface actionable information.

   Consider capturing and surfacing the test output in the failure message:

   ```js
   check(
     "Tests pass",
     () => {
       try {
         execFileSync("npm", ["test"], { stdio: "pipe" });
         return true;
       } catch (err) {
         const output = (err.stdout || "") + (err.stderr || "");
         console.error("\nTest output:\n" + output.slice(0, 2000));
         return false;
       }
     },
     "Tests are failing. Fix them before shipping: npm test",
   );
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge-preflight.js
   Line: 219-224

   Comment:
   **`npm test` output silently discarded on failure**

   `execFileSync("npm", ["test"], { stdio: "pipe" })` captures and then discards all stdout/stderr from the test run. When this check fails, `forge-preflight ship` reports only:

   ```
   ✗ Tests pass
       Tests are failing. Fix them before shipping: npm test
   ```

   The user gets no indication of _which_ tests failed or _why_ — they must re-run `npm test` manually to see the output. This defeats part of the purpose of a preflight check that's meant to surface actionable information.

   Consider capturing and surfacing the test output in the failure message:

   ```js
   check(
     "Tests pass",
     () => {
       try {
         execFileSync("npm", ["test"], { stdio: "pipe" });
         return true;
       } catch (err) {
         const output = (err.stdout || "") + (err.stderr || "");
         console.error("\nTest output:\n" + output.slice(0, 2000));
         return false;
       }
     },
     "Tests are failing. Fix them before shipping: npm test",
   );
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

4. `bin/forge-preflight.js`, line 18 ([link](https://github.com/harshanandak/forge/blob/2bcee14f0a7391e0b7b4bcc5ef72aa97e65ffdfb/bin/forge-preflight.js#L18)) 

   <a href="#">P2</a> **Dead commented-out import**

   `path` is imported but commented out and never used in this file. It should be removed to keep the file clean.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge-preflight.js
   Line: 18

   Comment:
   **Dead commented-out import**

   `path` is imported but commented out and never used in this file. It should be removed to keep the file clean.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

5. `bin/forge-preflight.js`, line 117-121 ([link](https://github.com/harshanandak/forge/blob/823219b1f947df4987f57e08422ccc48bade0bdd/bin/forge-preflight.js#L117-L121)) 

   <a href="#">P1</a> **Branch prefix allowlist inconsistent with generic slug stripping in `smart-status.sh`**

   `validateDev` hard-codes only `feat/`, `fix/`, and `docs/` as valid feature-branch prefixes. With this PR, `smart-status.sh` was simultaneously updated to strip **any** prefix generically (`_slug="${_slug##*/}"`), so it now correctly detects sessions for branches like `chore/update-deps`, `refactor/auth-flow`, `hotfix/null-check`, `bugfix/crash`, etc.

   A developer who creates such a branch — which `smart-status` happily links to an in-progress issue — will then be blocked by `forge-preflight dev` with:

   ```
   ✗ On feature branch
       Not on a feature branch. Create one: git checkout -b feat/your-feature
   ```

   This makes the two tools contradict each other on what constitutes an active development branch.

   

   Alternatively, a single check for the presence of a `/` (mirroring the `*/*` pattern in `smart-status.sh`) would future-proof this without enumerating every possible prefix:

   ```js
   return branch.includes("/") && branch !== "HEAD";
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge-preflight.js
   Line: 117-121

   Comment:
   **Branch prefix allowlist inconsistent with generic slug stripping in `smart-status.sh`**

   `validateDev` hard-codes only `feat/`, `fix/`, and `docs/` as valid feature-branch prefixes. With this PR, `smart-status.sh` was simultaneously updated to strip **any** prefix generically (`_slug="${_slug##*/}"`), so it now correctly detects sessions for branches like `chore/update-deps`, `refactor/auth-flow`, `hotfix/null-check`, `bugfix/crash`, etc.

   A developer who creates such a branch — which `smart-status` happily links to an in-progress issue — will then be blocked by `forge-preflight dev` with:

   ```
   ✗ On feature branch
       Not on a feature branch. Create one: git checkout -b feat/your-feature
   ```

   This makes the two tools contradict each other on what constitutes an active development branch.

   

   Alternatively, a single check for the presence of a `/` (mirroring the `*/*` pattern in `smart-status.sh`) would future-proof this without enumerating every possible prefix:

   ```js
   return branch.includes("/") && branch !== "HEAD";
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: bin/forge-preflight.js
Line: 117-121

Comment:
**Branch prefix allowlist inconsistent with generic slug stripping in `smart-status.sh`**

`validateDev` hard-codes only `feat/`, `fix/`, and `docs/` as valid feature-branch prefixes. With this PR, `smart-status.sh` was simultaneously updated to strip **any** prefix generically (`_slug="${_slug##*/}"`), so it now correctly detects sessions for branches like `chore/update-deps`, `refactor/auth-flow`, `hotfix/null-check`, `bugfix/crash`, etc.

A developer who creates such a branch — which `smart-status` happily links to an in-progress issue — will then be blocked by `forge-preflight dev` with:

```
✗ On feature branch
    Not on a feature branch. Create one: git checkout -b feat/your-feature
```

This makes the two tools contradict each other on what constitutes an active development branch.

```suggestion
        return (
          branch.startsWith("feat/") ||
          branch.startsWith("fix/") ||
          branch.startsWith("docs/") ||
          branch.startsWith("chore/") ||
          branch.startsWith("refactor/") ||
          branch.startsWith("hotfix/") ||
          branch.startsWith("bugfix/") ||
          branch.startsWith("test/") ||
          branch.startsWith("release/")
        );
```

Alternatively, a single check for the presence of a `/` (mirroring the `*/*` pattern in `smart-status.sh`) would future-proof this without enumerating every possible prefix:

```js
return branch.includes("/") && branch !== "HEAD";
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ["fix: reset checks ar..."](https://github.com/harshanandak/forge/commit/823219b1f947df4987f57e08422ccc48bade0bdd)</sub>

### CODE REVIEW

File: scripts/smart-status.sh [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        _remaining_branches="${_remaining_branches#*|}"
+        ;;
+      *)
+        _branch="$_remaining_branches"
+        _remaining_branches=""
+        ;;
+    esac
+
+    # Get changed files for this branch vs master (-- prevents injection)
+    _files="$("$GIT" diff "master...${_branch}" --name-only -- 2>/dev/null || echo '')"
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Auto-detect default branch (master/main) instead of hardcoding master.

Changed: Added BASE_BRANCH auto-detection via git rev-parse --verify, with DEFAULT_BRANCH env override
Reason: Repos using main instead of master would get empty conflict detection results
Commit: 8dc5aeb

File: scripts/smart-status.sh [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+        # With --name-only, conflicted files are listed after the SHA line
+        _conflict_files=""
+        _line_num=0
+        while IFS= read -r _cf_line; do
+          _line_num=$((_line_num + 1))
+          # Skip the first line (tree SHA) if present
+          case "$_cf_line" in
+            [0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]*) continue ;;
+            "") continue ;;
+          esac
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: SHA skip pattern now uses line number (skip line 1) instead of hex regex.

Changed: Replaced case pattern [0-9a-f]{7}* with line_num == 1 check
Reason: Hex regex could match legitimate filenames starting with hex chars
Commit: 8dc5aeb

File: scripts/smart-status.sh [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+EOF
+
+# Get in-progress issues for matching (only if we have sessions)
+IN_PROGRESS_JSON="[]"
+if [ "$SESSION_COUNT" -gt 0 ]; then
+  IN_PROGRESS_JSON="$("$BD" list --status in_progress --json --limit 0 2>/dev/null || echo '[]')"
+  # Fallback: if bd doesn't support --status flag, filter from full list
+  if [ "$(printf '%s' "$IN_PROGRESS_JSON" | jq 'length' 2>/dev/null)" = "0" ] || [ -z "$IN_PROGRESS_JSON" ]; then
+    IN_PROGRESS_JSON="$(printf '%s' "$ISSUES_JSON" | jq '[.[] | select(.status == "in_progress")]')"
+  fi
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Fallback now only triggers when bd command fails (empty/invalid output), not when result is legitimately empty array.

Changed: Check for empty string or invalid JSON instead of length==0
Reason: length==0 triggered fallback even with 0 in-progress issues, masking the real result
Commit: 8dc5aeb

File: scripts/smart-status.sh [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+      # Tier 1: Conflict risk annotations (file overlap only, excluding merge conflicts)
+      (if ($conflicts // [] | length) > 0 then
+        # Build set of files that have real merge conflicts
+        ([$merge_conflicts // [] | .[] | .files[]] | unique) as $mc_files |
+        ($conflicts[] |
+          .branch as $other |
+          .files[] |
+          # Only show Conflict risk if NOT already a merge conflict
+          select([$mc_files[] | select(. == .)] | length == 0) // . |
+          if ($mc_files | index(.)) then empty
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh:541]: Fixed: Removed redundant no-op filter select(. == .) which always matched.

Changed: Removed the line entirely, keeping the existing per-file membership check via index(.)
Reason: The select was a no-op that added noise without filtering anything
Commit: 8dc5aeb

File: scripts/smart-status.sh [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+          else "    ! Conflict risk: " + . + " (" + $other + " in-progress)"
+          end
+        )
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh:544]: Fixed: SHA skip pattern replaced with line-number check, and redundant no-op jq filter removed.

Changed: Line 1 skip via line_num == 1 instead of hex regex; removed select(. == .) no-op
Reason: Hex regex matched legitimate filenames; no-op filter added noise
Commit: 8dc5aeb

File: scripts/smart-status.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Auto-detect default branch (master/main) instead of hardcoding master.

Changed: Added BASE_BRANCH auto-detection via git rev-parse --verify with DEFAULT_BRANCH env override
Reason: Repos using main without a master ref would get empty conflict detection results
Commit: 8dc5aeb

File: package.json [hunk 6]
```diff
@@ -4,7 +4,7 @@
   "description": "7-stage TDD workflow for ALL AI coding agents (Claude, Cursor, Cline, OpenCode, Copilot, Kilo Code, Roo Code, Codex)",
   "bin": {
     "forge": "bin/forge.js",
-    "forge-validate": "bin/forge-validate.js"
+    "forge-preflight": "bin/forge-preflight.js"
   },
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply package.json:8]: Fixed: Remaining forge-validate references in docs/research/ updated to forge-preflight.

Changed: docs/research/pr4-cli-automation.md and docs/research/test-environment.md updated
Reason: After the rename, these docs still referenced the old binary name
Commit: 8dc5aeb

File: scripts/smart-status.sh [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+      '{sessions: $sessions, issues: $issues}'
+  else
+    printf '%s\n' "$SCORED_JSON"
+  fi
+else
+  # Grouped, colored output
+  # NO_COLOR support: https://no-color.org/
+  if [ -n "${NO_COLOR:-}" ]; then
+    C_RESET="" C_GREEN="" C_YELLOW="" C_RED="" C_DIM="" C_BOLD="" C_CYAN=""
+  else
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: JSON output now always returns consistent shape {sessions: [], issues: [...]} regardless of session count.

Changed: Removed conditional branching in --json output; always wraps with sessions+issues
Reason: Consumers shouldn't need to check whether output is array or object
Commit: cf9fa05

File: test/scripts/smart-status.test.js [hunk 8]
```diff
@@ -0,0 +1,1339 @@
...
+`;
+  fs.writeFileSync(mockScript, scriptContent, { mode: 0o755 });
+  return { tmpDir, mockScript };
+}
+
+function cleanupTmpDir(tmpDir) {
+  try {
+    fs.rmSync(tmpDir, { recursive: true, force: true });
+  } catch (_e) {
+    // ignore cleanup errors
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply test/scripts/smart-status.test.js:113]: The mock bd does handle --status in_progress via the fallback path in smart-status.sh. Session matching is tested in 7 dedicated tests (describe 'session detection') covering: branch-to-issue matching, orphan branches, no-worktree skip, JSON sessions array, and multiple worktrees with mixed matching.

The mock returns all issues and the script filters in_progress internally via jq. This mirrors real behavior since bd list --status may not be supported on all versions.

See: test/scripts/smart-status.test.js lines 760-870 (session detection tests)

File: scripts/smart-status.sh [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+  case "$line" in
+    worktree\ *)
+      _wt_path="${line#worktree }"
+      ;;
+    branch\ *)
+      _wt_branch="${line#branch refs/heads/}"
+      ;;
+    "")
+      # End of block — process if we have a branch that is not main/master
+      if [ -n "$_wt_branch" ] && [ "$_wt_branch" != "main" ] && [ "$_wt_branch" != "master" ]; then
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Session filter now uses BASE_BRANCH (auto-detected) instead of hardcoded main/master. Also keeps main and master as fallbacks for safety.

Changed: Line 217 now checks _wt_branch != BASE_BRANCH in addition to != main and != master
Reason: Custom default branches (e.g. develop, trunk) would appear as active sessions
Commit: ee9f358

File: scripts/smart-status.sh [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+
+    # Get changed files for this branch vs master (-- prevents injection)
+    _files="$("$GIT" diff "${BASE_BRANCH}...${_branch}" --name-only -- 2>/dev/null || echo '')"
+    # Collapse to comma-delimited, strip empty lines
+    _files_csv="$(printf '%s' "$_files" | tr '\n' ',' | sed 's/,$//' | sed 's/^,//')"
+
+    if [ -n "$ALL_BRANCH_FILES" ]; then
+      ALL_BRANCH_FILES="${ALL_BRANCH_FILES}|${_branch}:${_files_csv}"
+    else
+      ALL_BRANCH_FILES="${_branch}:${_files_csv}"
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: File lists now use tab delimiter instead of comma. Tabs cannot appear in filenames on any major filesystem.

Changed: tr '\n' ',' replaced with tr '\n' '\t' throughout; jq split(",") replaced with split("\t")
Reason: Filenames with commas would corrupt the file list and break conflict detection
Commit: ee9f358

File: scripts/smart-status.sh [hunk 11]
```diff
@@ -0,0 +1,645 @@
...
+
+# Bail early on empty
+if [ "$(printf '%s' "$ISSUES_JSON" | jq 'length')" = "0" ]; then
+  if [ "$JSON_MODE" = "1" ]; then
+    echo "[]"
+  else
+    echo "No issues found."
+  fi
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh:103]: Fixed: Early-exit now returns consistent {sessions:[], issues:[]} envelope instead of bare []. Commit: 63dfbba

File: test/scripts/smart-status.test.js [hunk 12]
```diff
@@ -0,0 +1,1357 @@
...
+        expect(result.status).toBe(0);
+        expect(result.stdout).toContain('ACTIVE SESSIONS');
+        expect(result.stdout).toContain('feat/alpha');
+        expect(result.stdout).toContain('Changed:');
+        expect(result.stdout).toContain('src/a.js');
+        expect(result.stdout).toContain('src/b.js');
+        expect(result.stdout).toContain('feat/beta');
+        expect(result.stdout).toContain('src/c.js');
+      } finally {
+        cleanupTmpDir(mockBd.tmpDir);
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply test/scripts/smart-status.test.js:973]: Fixed: parseIssues now enforces {sessions, issues} envelope and throws if missing. Early-exit also returns consistent envelope. Commit: 63dfbba

File: scripts/smart-status.sh [hunk 13]
```diff
@@ -0,0 +1,645 @@
...
+        # Status + days
+        ("[" + ($item.status // "open") + " " + ($item.days_ago | tostring) + "d]") as $status_tag |
+        # Unblocks annotation
+        (if ($item.dependents // [] | length) > 0 then
+          "\n   -> Unblocks: " + ($item.dependents | join(", "))
+         else "" end) as $unblocks |
+        "ENTRY:" + $item.group + ":" +
+          $rank + ". [" + ($item.score | tostring) + "] " +
+          $item.id + " (" + ($item.priority // "-") + " " + ($item.type // "-") + ") -- " +
+          ($item.title // "-") + " " + $status_tag + $stale + $unblocks
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Removed embedded newline from unblocks annotation. Now uses inline ' -> Unblocks:' instead of newline prefix, so the while-read loop processes entries correctly. Commit: 63dfbba

**Comment by harshanandak** [comment]
@greptileai /review

File: scripts/smart-status.sh [hunk 14]
```diff
@@ -0,0 +1,645 @@
...
+SESSIONS_JSON="[]"
+if [ "$SESSION_COUNT" -gt 0 ]; then
+  # Process each session (newline-delimited "branch<TAB>path" entries)
+  while IFS='	' read -r _branch _path; do
+    [ -z "$_branch" ] && continue
+
+    # Extract slug from branch name (strip any <prefix>/ convention)
+    _slug="$_branch"
+    case "$_slug" in
+      */*) _slug="${_slug#*/}" ;;
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh]: Fixed: Changed slug stripping from #*/ (shortest match) to ##*/ (greedy match). Now team/feat/my-feature -> my-feature correctly. Commit: b7f1d9f

File: scripts/smart-status.sh [hunk 15]
```diff
@@ -0,0 +1,645 @@
...
+    # Unblock chain: dependent_count + 1 (min 1)
+    (((.dependent_count // 0) + 1) | if . < 1 then 1 else . end) as $unblock_chain |
+
+    # Type weight: bug=1.2, feature=1.0, task=0.8, default=1.0
+    (if .type == "bug" then 1.2
+     elif .type == "feature" then 1.0
+     elif .type == "task" then 0.8
+     else 1.0 end) as $type_weight |
+
+    # Status boost: in_progress=1.5, open=1.0, default=1.0
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh:158]: Added clarifying comment explaining the intentional divergence: dependent_count is the authoritative server-side count used for scoring, while the Unblocks annotation uses the local response which may show fewer items when bd list omits closed/filtered issues. Commit: b7f1d9f

File: scripts/smart-status.sh [hunk 16]
```diff
@@ -0,0 +1,648 @@
...
+        ((.updated_at | sub("\\.[0-9]+Z$"; "Z")) | fromdateiso8601) as $ts |
+        (($now | tonumber) - $ts) / 86400 | floor
+       else 0 end) as $days_ago |
+      # Group assignment (priority order: RESUME > UNBLOCK_CHAINS > BLOCKED > BACKLOG > READY_WORK)
+      (if .status == "in_progress" then "RESUME"
+       elif (.dependent_count // 0) >= 2 and .status != "in_progress" then "UNBLOCK_CHAINS"
+       elif (.dependency_count // 0) > 0 and .status != "closed" then "BLOCKED"
+       elif .priority == "P4" then "BACKLOG"
+       else "READY_WORK"
+       end) as $group |
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/smart-status.sh:567]: Fixed: in_progress issues with active blockers now stay in RESUME (since work has started) but get a '\!\! BLOCKED by dependencies' warning annotation. Added blocked_resume flag to the JSON output. Also added clarifying comment explaining the design decision. Commit: 7c74491

File: bin/forge-preflight.js [hunk 17]
```diff
@@ -1,22 +1,21 @@
...
  *
  * Usage:
- *   forge-validate status  - Check project prerequisites
- *   forge-validate dev     - Validate before /dev stage
- *   forge-validate ship    - Validate before /ship stage
+ *   forge-preflight status  - Check project prerequisites
+ *   forge-preflight dev     - Validate before /dev stage
+ *   forge-preflight ship    - Validate before /ship stage
  *
  * Security: Uses execFileSync to prevent command injection.
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge-preflight.js:14]: Fixed: Added checks = [] at the start of validateStatus(), validateDev(), and validateShip() to prevent check accumulation across sequential calls. Commit: 823219b
